web/modules/custom/rdf_entity_provenance/src/ProvenanceHelper.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\rdf_entity_provenance;
+
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\rdf_entity\RdfInterface;
+
+/**
+ * Provides helper methods to fetch, check and update provenance data.
+ */
+class ProvenanceHelper implements ProvenanceHelperInterface {
+
+  /**
+   * The entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * Constructs the ProvenanceHelper service object.
+   *
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entityTypeManager
+   *   The entity type manager.
+   */
+  public function __construct(EntityTypeManagerInterface $entityTypeManager) {
+    $this->entityTypeManager = $entityTypeManager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function isFederated(RdfInterface $rdf_entity): bool {
+    return empty($this->getProvenanceActivity($rdf_entity));
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getProvenanceActivity(RdfInterface $rdf_entity): RdfInterface {
+    if (!$activity = $this->loadProvenanceActivity($rdf_entity)) {
+      $activity = $this->createProvenanceActivity($rdf_entity);
+    }
+
+    return $activity;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function loadProvenanceActivity(RdfInterface $rdf_entity): ?RdfInterface {
+    /** @var \Drupal\rdf_entity\RdfInterface $activity */
+    $activity = $this->getStorage()->loadByProperties([
+      'bundle' => 'provenance_activity',
```

Comment:
s/bundle/rid ?

Reply:
And also, `:: loadByProperties()` returns an array. We should use `reset()` or `array_shift()` to pickup the first item and ensure a NULL if the array is empty.